README.md
# Notice

The old clojars `kaiuri/yahoo-finance` has been moved to `kaiuri/clj-yahoo-finance`.

# clj-yahoo-finance

[![Clojars Project](https://img.shields.io/clojars/v/net.clojars.kaiuri/clj-yahoo-finance.svg)](https://clojars.org/net.clojars.kaiuri/clj-yahoo-finance)

Clojure library for fetching [_yahoo!finance_](finance.yahoo.com) historical data. No HTML parsing.

## Usage

`clj-yahoo-finance` provides a function `get-historical` which takes dict:

- `stock`: Stock's name (string) as displayed in `finance.yahoo.com`.
- `period`: PersistentVector tuple of date strings, each in the format `YYYY-MM-dd`.
- `[frequency]`: String, one of `w`(week), `d`(day) or `m`(month).
- `[extension]`: One of `:clojure`, `:csv`, `:json` or `:yaml`, defaults to `:clojure`.

And returns the data.

```clojure
(require '[kaiuri.clj-yahoo-finance.core :as yf])
(->> (yf/get-historical {:stock "^RUT", :period ["2021-10-10" "2022-01-01"]
                         :frequency "d" :extension :clojure})
     (println))

#_=> [{:date 2021-10-11, :open 2233.000000, :high 2251.649902, :low 2220.639893, :close 2220.639893, :adj-close 2220.639893, :volume 2580000000}, ...]
```

### Extras

Tiny parsers are exposed at [parsers.clj](./src/kaiuri/clj_yahoo_finance/parsers.clj), those are subject to changes but work pretty well.
Current available parsers.

- `csv->yaml`,
- `csv->clj`,
- `csv->json`.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [1.0.15] - 17-07-22

- Add `json` and `yaml` output formats.
- Refactor parsers into their own namespace.
- Improved spec validation.
- More idiomatic Clojure code.


test/kaiuri/yahoo_finance_test.clj
(ns kaiuri.yahoo-finance-test
  (:require
    [clojure.test :as t :refer [deftest is testing]]
    [kaiuri.clj-yahoo-finance.core :as yf]))

(deftest get-historical-clojure
  (testing "kaiuri.clj-yahoo-finance.core/get-historical"
    (let [query {:stock "CL=F"
                 :period ["2022-01-01" "2022-02-01"]
                 :frequency "d"
                 :extension :clojure}
          data (yf/get-historical query)]
      (is some? data)
      (is some? (take 2 data)))))

(deftest get-historical-csv
  (testing "kaiuri.clj-yahoo-finance.core/get-historical"
    (let [query {:stock "^IXIC"
                 :period ["2022-01-01" "2022-02-01"]
                 :frequency "d"
                 :extension :csv}
          data (yf/get-historical query)]
      (is some? data)
      (is some? (take 2 data)))))

(deftest get-historical-yaml
  (testing "kaiuri.clj-yahoo-finance.core/get-historical"
    (let [query {:stock "^RUT"
                 :period ["2022-01-01" "2022-02-01"]
                 :frequency "d"
                 :extension :yaml}
          data (yf/get-historical query)]
      (is some? data)
      (is some? (take 2 data)))))

(deftest get-historical-json
  (testing "kaiuri.clj-yahoo-finance.core/get-historical"
    (let [query {:stock "^RUT"
                 :period ["2022-01-01" "2022-02-01"]
                 :frequency "d"
                 :extension :json}
          data (yf/get-historical query)]
      (is some? data)
      (is some? (take 2 data)))))


build.clj
(ns build
  (:refer-clojure :exclude [test])
  (:require [clojure.tools.build.api :as b] ; for b/git-count-revs
            [org.corfield.build :as bb]))

(def lib 'net.clojars.kaiuri/clj-yahoo-finance)
; (def version "0.1.0-SNAPSHOT")
; alternatively, use MAJOR.MINOR.COMMITS:
(def version (format "1.0.%s" (b/git-count-revs nil)))

(defn test "Run the tests." [opts]
  (bb/run-tests opts))

(defn ci "Run the CI pipeline of tests (and build the JAR)." [opts]
  (-> opts
      (assoc :lib lib :version version)
      (bb/run-tests)
      (bb/clean)
      (bb/jar)))

(defn install "Install the JAR locally." [opts]
  (-> opts
      (assoc :lib lib :version version)
      (bb/install)))

(defn deploy "Deploy the JAR to Clojars." [opts]
  (-> opts
      (assoc :lib lib :version version)
      (bb/deploy)))


src/kaiuri/clj_yahoo_finance/core.clj
(ns kaiuri.clj-yahoo-finance.core
  (:require
    [clojure.spec.alpha :as s]
    [kaiuri.clj-yahoo-finance.parsers
     :refer [csv->clj csv->yaml csv->json]]
    [kaiuri.clj-yahoo-finance.specs]
    [kaiuri.clj-yahoo-finance.utils :refer [period->url]])
  (:import
    [java.net URLEncoder]))

(set! *warn-on-reflection* true)

(defn- query1-v7
  "PRIVATE: Uses the current (12-07-22) finance.yahoo.com routing to get historical data."
  [{:keys [stock
           ^clojure.lang.PersistentVector period
           frequency]}]
  (let [stock-chunk (URLEncoder/encode stock)
        period-url (period->url period)
        frequency-chunk (str \1 frequency)]
    (str "https://query1.finance.yahoo.com/v7/finance/download/" stock-chunk
         \? period-url
         "&interval=" frequency-chunk
         "&events=history"
         "&includeAdjustedClose=true")))

(defn get-historical
  "Takes a dict with the following keys:
    - <stock>: Stock's name (string) as displayed in `finance.yahoo.com`.
    - <period>: PersistentVector tuple of date strings, each in the format `YYYY-MM-dd`.
    - [frequency]: `w`(week), `d`(day) or `m`(month).
    - [extension]: One of :clojure|:csv|:yaml|:json, defaults to `:clojure`.

   Returns historical data in the specified extension.
   "
  [{:keys [stock
           ^clojure.lang.PersistentVector period
           frequency
           extension]
    :or {frequency "d" extension :clojure}}]

  {:pre [(s/valid? :query/period period)
         (s/valid? :query/frequency frequency)
         (s/valid? :query/extension extension)
         (s/valid? :query/stock stock)]}

  (let [raw-data (->> {:stock stock
                       :period period
                       :frequency frequency}
                      (query1-v7)
                      (slurp))]
    (case extension
      :csv raw-data
      :clojure (csv->clj raw-data)
      :json (csv->json raw-data)
      :yaml (csv->yaml raw-data))))


src/kaiuri/clj_yahoo_finance/specs.clj
(ns kaiuri.clj-yahoo-finance.specs
  (:require [clojure.spec.alpha :as spec]))

(defn time?
  [s]
  (and (string? s)
       (re-matches #"^\d{4}(?:\D\d{2}){2}" s)))

(spec/def :date/format
  (spec/and string? #(re-matches #"^\d{4}(?:\D\d{2}){2}" %)))

(spec/def :query/period
  (spec/coll-of :date/format :kind vector? :count 2 :distinct true))

(spec/def :query/stock string?)

(spec/def :query/frequency
  #(re-matches #"d|w|m" (str %)))

(spec/def :query/extension
  (spec/or :yaml #(re-matches #"yaml" (name %))
           :csv #(re-matches #"csv" (name %))
           :clojure #(re-matches #"clojure" (name %))
           :json #(re-matches #"json" (name %))))

(spec/def :query/query
  (spec/keys :req-un [:query/stock :query/period  :query/extension]
             :opt-un [:query/frequency :query/extension]))


src/kaiuri/clj_yahoo_finance/utils.clj
(ns kaiuri.clj-yahoo-finance.utils
  (:require
    [clojure.instant :as instant]
    [clojure.spec.alpha :as spec]
    [clojure.string :as string]
    [kaiuri.clj-yahoo-finance.specs])
  (:import [java.util TimeZone
            GregorianCalendar]))

(set! *warn-on-reflection* true)

(defn- gregorian->unix
  [date]
  {:pre [(spec/valid? :date/format date)]}
  (let [^TimeZone timezone (TimeZone/getTimeZone "America/New_York")
        ^GregorianCalendar calendar (instant/read-instant-calendar date)]
    (doto calendar
      (.setTimeZone timezone))
    (quot (.getTimeInMillis calendar) 1000)))

(defn- index-of
  [^clojure.lang.PersistentVector v element & {:keys [offset] :or {offset 0}}]
  (+ offset (.indexOf v element)))

(defn period->url
  [^clojure.lang.PersistentVector period]
  {:pre [(spec/valid? :query/period period)]}
  (->> period
       (map #(str "period" (index-of period % :offset 1) \= (gregorian->unix %)))
       (string/join \&)))


src/kaiuri/clj_yahoo_finance/parsers.clj
(ns kaiuri.clj-yahoo-finance.parsers
  (:require [clojure.string :as string]
            [camel-snake-kebab.core :refer [->snake_case ->camelCase ->kebab-case]]
            [clj-yaml.core :as yaml]
            [cheshire.core :as json]))

(defn csv->clj
  "PRIVATE: Subject to changes.
   Parses CSV string into vector of dicts.
    - <s>: CSV string.
    - [del]: Delimiter, defaults to `,`
  "
  [s & {:keys [del] :or {del #","}}]
  (let [csv (string/split-lines s)
        head (map #(keyword (->kebab-case %)) (string/split (first csv) del))
        body (map #(string/split % del) (rest csv))]
    (mapv #(zipmap head %) body)))

(defn csv->json
  "PRIVATE: Subject to changes.
   Parses CSV string into json string.
    - <s>: CSV string.
    - [del]: Delimiter, defaults to `,`
  "
  [s & {:keys [del] :or {del #","}}]
  (let [csv (string/split-lines s)
        head (map #(keyword (->camelCase %)) (string/split (first csv) del))
        body (map #(string/split % del) (rest csv))]
    (json/generate-string (mapv #(zipmap head %) body))))

(defn csv->yaml
  "PRIVATE: Subject to changes.
   Parses CSV string into yaml string.
    - <s>: CSV string.
    - [del]: Delimiter, defaults to `,`
  "
  [s & {:keys [del] :or {del #","}}]
  (let [csv (string/split-lines s)
        head (map #(keyword (->snake_case %)) (string/split (first csv) del))
        body (map #(string/split % del) (rest csv))]
    (yaml/generate-string (mapv #(zipmap head %) body))))


src/kaiuri/yahoo_finance.clj
(ns kaiuri.yahoo-finance
  (:require [clojure.string :as string]
            [camel-snake-kebab.core :as csk])
  (:import [java.net URLEncoder]
           [java.time LocalDate ZoneId]))

(set! *warn-on-reflection* true)

(defn- str->milli
  [s]
  (let [date (LocalDate/parse s)
        zone (ZoneId/systemDefault)
        milli (-> date
                  (.atStartOfDay zone)
                  .toInstant
                  .toEpochMilli)]
    (quot milli 1000)))

(defn- query1-v7
  "PRIVATE: Uses the current (12-07-22) finance.yahoo.com routing to get historical data."
  [{:keys [stock
           ^clojure.lang.PersistentVector period
           frequency]}]
  (let [stock-chunk (URLEncoder/encode stock)
        period-chunk (->> period
                          (map (fn [p]
                                 (let [idx (+ 1 (.indexOf period p))]
                                   (str "period" idx \= (str->milli p)))))
                          (string/join \&))
        frequency-chunk (str \1 frequency)]
    (str "https://query1.finance.yahoo.com/v7/finance/download/" stock-chunk
         \? period-chunk
         "&interval=" frequency-chunk
         "&events=history"
         "&includeAdjustedClose=true")))

(defn- csv->clj
  "PRIVATE: Subject to changes.
   Parses CSV string into vector of dicts.
    - <s>: CSV string.
    - [del]: Delimiter, defaults to `,`

  "
  [s & {:keys [del] :or {del #","}}]
  (let [csv (string/split-lines s)
        head (map #(keyword (csk/->kebab-case %)) (string/split (first csv) del))
        body (map #(string/split % del) (rest csv))]
    (mapv #(zipmap head %) body)))

(defn get-historical
  "Takes a dict with the following keys:
    - <stock>: Stock's name (string) as displayed in `finance.yahoo.com`.
    - <period>: PersistentVector tuple of date strings, each in the format `YYYY-MM-dd`.
    - [frequency]: `w`(week), `d`(day) or `m`(month).
    - [extension]: One of `:clojure`, `:raw`, defaults to `:clojure`.

   Returns historical data in the specified extension.
   "

  [{:keys [stock
           ^clojure.lang.PersistentVector period
           frequency
           extension]
    :or {frequency "d" extension :clojure}}]
  (let [raw-data (->> {:stock stock
                       :period period
                       :frequency frequency}
                      (query1-v7)
                      (slurp))]
    (case extension
      :raw raw-data
      :clojure (csv->clj raw-data)
      :else (csv->clj raw-data))))
